## CT4101 - Deep Learning CT5133
### Student Name(s): Tapan Auti, Atharva Kulkarni
### Student ID(s): 20231499, 20231773

# Part 1

Neural Network Approach to Logistic Regression:

Logistic Regression is a Classification method where the goal of the algorithm is to find a dividing line between classes. This is contrary to Linear Regression where the goal of the algorithm is to find a line that goes through all the data points.

Using a Neural Network Approach to Logistic Regression would be as follows:

Logistic Regression is equivalent to a one layer deep neural network

1. We have a single neuron that recieves an input with a weight and a bias.
2. This neuron has an activation function (suppose a sigmoid/logit function) which activates when the weighted sum is above a paricular threshold

Description of Algorithm:



Blobs Dataset - Linearly Separable

Moons Dataset - Non Separable

References:

1. Prof M.Madden - Lecture Notes, CT5133 - Deep Learning
2. [Towards Data Science - Intro To Logistic Regression](https://towardsdatascience.com/introduction-to-logistic-regression-66248243c148#:~:text=Logistic%20Regression%20is%20a%20Machine,on%20the%20concept%20of%20probability.&text=The%20hypothesis%20of%20logistic%20regression,function%20between%200%20and%201%20.)
3. [Logistic Regression Using a Neural Network Mindset](https://www.youtube.com/watch?v=rGoJAL-rkE0&ab_channel=TechwithKumar)
4. [But What is a Neural Network?](https://www.youtube.com/watch?v=aircAruvnKk&ab_channel=3Blue1Brown)
5. [Logistic Regression](https://www.youtube.com/watch?v=zM4VZR0px8E&ab_channel=codebasics)
6. [Deep Learning, Logistic Regression with a Neural Network Mindset](https://github.com/Kulbear/deep-learning-coursera/blob/master/Neural%20Networks%20and%20Deep%20Learning/Logistic%20Regression%20with%20a%20Neural%20Network%20mindset.ipynb)
7. [Dummy Variable and One Hot Encoding](https://www.youtube.com/watch?v=9yl6-HEY7_s&ab_channel=codebasics)
8. [Logistic Regression using a Neural Network Approach](https://datascience-enthusiast.com/DL/Logistic-Regression-with-a-Neural-Network-mindset.html)
9. [Log Regression NN Mindset](https://edorado93.github.io/2018/09/07/Logistic-Regression-with-a-Neural-Networks-Mindset-9b5526c2ed46/)

In [1]:
import matplotlib
import matplotlib.pyplot as plt
import sklearn
import sklearn.datasets
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

# Display plots inline and change default figure size
%matplotlib inline

# 1 Author: Atharva
# Sigmoid Function Definition


def f(z):
    return 1 / (1 + np.exp(-z))


# 1 Author: Tapan
# Calculate loss function
def loss_gradient(w, x, b):
    z = np.dot(x, w)
    y = f(z)
    loss = np.mean(-b * np.log(y) - (1 - b) * np.log(1 - y))
    gdn = np.dot(x.T, (b - y)) / b.size
    # print(b.shape)
    # print(y.shape)

    # 1 Author: Atharva
    # Normalising Gradient Descent
    grad = gdn + 300 / (2 * b.size) * np.concatenate(([0], w[1:])).T
    return loss, grad


# 1 Author: Tapan
# Training Function
def model_fit(x, b, total_iterations=5000):
    final_w = []
    label = np.unique(b)
    all_loss = np.zeros(total_iterations)

    for c in label:
        b = np.where(b == c, 1, 0)
        w = np.zeros(x.shape[1])
        for i in range(total_iterations):
            all_loss[i], grad = loss_gradient(w, x, b)
            w += 0.01 * grad
        final_w.append(w)
    return final_w, label, all_loss


# 1 Author: Atharva
# Predict Function
def predict_prob(label, w, x):
    preds = [np.argmax([f(i@j) for j in final_w]) for i in x]
    return [label[p] for p in preds]


# 1 Author: Atharva
# Calculate Accuracb of the model
def accuracy(label, w, x, b):
    acc = predict_prob(label, w, x)
    return (acc == b).mean()


# Part 2

References:

1. But What is A

In [2]:
df = pd.read_csv("moons400.csv") # Moons Dataset
df2 = pd.read_csv("blobs250.csv") # Blobs Dataset


# y = df['Class'].values
# y2 = df2['Class'].values

# # Drop the 'Class' Columns
# del df['Class']
# del df2['Class']

# X = df.values   
# X2 = df2.values

# Test the model on Moons Dataset
data_moons = np.array(df)

num_train = int(.70 * len(data_moons))
num_test = int(0.15 * len(data_moons))

x_train, y_train = data_moons[:num_train, :-1], data_moons[:num_train, -1]
x_test, y_test = data_moons[num_test:, :-1], data_moons[num_test:, -1]

final_w, label, losses = model_fit(x_train, y_train)
print(f"Test Accuracy Moons  : {100*accuracy(label, final_w, x_test, y_test):.2f}%")



FileNotFoundError: [Errno 2] No such file or directory: 'moons400.csv'

In [4]:
# Test the model on Blobs Dataset
data_blobs = np.array(df2)

num_train = int(.70 * len(data_blobs))
num_test = int(0.15 * len(data_blobs))

x_train, y_train = data_blobs[:num_train, :-1], data_blobs[:num_train, -1]
x_test, y_test = data_blobs[num_test:, :-1], data_blobs[num_test:, -1]

final_w, label, losses = model_fit(x_train, y_train)
print(f"Test Accuracy Blobs  : {100*accuracy(label, final_w, x_test, y_test):.2f}%")

Test Accuracy Blobs  : 49.30%


### Observations

1. The Moons dataset provides us an accuracy of 51.47%
2. The Blobs dataset providds us an accuracy of 49.3%
3. Even after being linearly separable, the Blobs dataset is predicted with an accuracy lower than that of the Moons dataset, when using Logistic Regression.

# Part 3

Description of Shallow Neural Network Algorithm:



References
1. [Shallow Neural Networks Explained](https://towardsdatascience.com/shallow-neural-networks-23594aa97a5#:~:text=Shallow%20neural%20networks%20consist%20of,inside%20a%20deep%20neural%20network.&text=The%20figure%20below%20shows%20a,layer%20and%201%20output%20layer.)
2. [Building](https://towardsdatascience.com/building-a-shallow-neural-network-a4e2728441e0)

In [5]:
def initialize_parameters(n_x, n_h, n_y):
    
    np.random.seed(2) 
    
    W1 = np.random.randn(n_h, n_x) * 0.01
    b1 = np.zeros(shape=(n_h, 1))
    W2 = np.random.randn(n_y, n_h) * 0.01
    b2 = np.zeros(shape=(n_y, 1))
    
    assert (W1.shape == (n_h, n_x))
    assert (b1.shape == (n_h, 1))
    assert (W2.shape == (n_y, n_h))
    assert (b2.shape == (n_y, 1))
    
    parameters = {"W1": W1,
                  "b1": b1,
                  "W2": W2,
                  "b2": b2}
    
    return parameters

def relu(x):
    return x*(x>0)
def sigmoid(x):
    return 1/(1+np.exp(-x))
def tanh(x):
    #return np.exp((np.exp(x)-np.exp(-x))-(np.exp(x)+np.exp(-x)))
    #return (np.exp(x)-np.exp(-x))/(np.exp(x)+np.exp(-x))
    return np.tanh(x)
    #return math.tanh(x)

def forward_pass(X, parameters):
    
    # to make forward pass calculations we need W1 and W2 so we will extract them from dictionary parameters
    W1 = parameters['W1']
    W2 = parameters['W2']
    b1 = parameters['b1']
    b2 = parameters['b2']

    # first layer calculations - hidden layer calculations
    Z1 = np.dot(W1,X) + b1
    A1 = tanh(Z1)  # activation in the first layer is tanh
    
    # output layer calculations
    Z2 = np.dot(W2, A1) + b2
    A2 = sigmoid(Z2)# A2 are predictions, y_hat
    
    # cache values for backpropagation calculations
    cache = {'Z1':Z1,
             'A1':A1,
             'Z2':Z2,
             'A2':A2
            }
    # print(A2.shape)
    return A2, cache

def backward_pass(parameters, cache, X, Y):
    
    # unpack paramaeters and cache to get values for backpropagation calculations
    W1 = parameters['W1']
    W2 = parameters['W2']
    
    Z1 = cache['Z1']
    A1 = cache['A1']
    Z2 = cache['Z2']
    A2 = cache['A2']
    
    m = X.shape[1] # number of examples in a training set
    #print(A2)
    #print(A2.shape)
    dZ2= A2 - Y
    
    
    dW2 = (1 / m) * np.dot(dZ2, A1.T)
    db2 = (1 / m) * np.sum(dZ2, axis=1, keepdims=True)   # keepdims - prevents python to output rank 1 array (n,)
    
    dZ1 = np.multiply(np.dot(W2.T, dZ2), 1 - np.power(A1, 2)) # we use tanh activation function
    dW1 = (1 / m) * np.dot(dZ1, X.T)  
    db1 = (1 / m) * np.sum(dZ1, axis=1, keepdims=True)
    
    grads = {"dW1": dW1,
             "db1": db1,
             "dW2": dW2,
             "db2": db2}
    
    return grads

def update_parameters(parameters, learning_rate, grads):
    
    W1 = parameters['W1']
    b1 = parameters['b1']
    W2 = parameters['W2']
    b2 = parameters['b2']
    
    dW1 = grads['dW1']
    db1 = grads['db1']
    dW2 = grads['dW2']
    db2 = grads['db2']
    
    W1 = W1 - learning_rate * dW1
    b1 = b1 - learning_rate * db1
    W2 = W2 - learning_rate * dW2
    b2 = b2 - learning_rate * db2
    
    # updated parameters
    parameters = {"W1": W1,
                  "b1": b1,
                  "W2": W2,
                  "b2": b2}
    
    return parameters

def NN_model(X,Y,n_h, num_iterations, learning_rate):
    
    n_x = X.shape[0] # size of an input layer = number of features 
    n_y = Y.shape[0] # size of an output layer
    parameters = initialize_parameters(n_x, n_h, n_y)
    
    #unpack parameters
    W1 = parameters["W1"]
    b1 = parameters["b1"]
    W2 = parameters["W2"]
    b2 = parameters["b2"]
    
    for i in range(num_iterations):
        A2, cache = forward_pass(X, parameters)
        grads = backward_pass(parameters, cache, X, Y)
        parameters = update_parameters(parameters, learning_rate, grads)
        
    return parameters

def predict(parameters, X):
    
    A2, cache = forward_pass(X, parameters)
    predictions = np.round(A2)
    
    return predictions

In [ ]:
data_moons = np.array(df)

num_train = int(.70 * len(data_moons))
num_test = int(0.15 * len(data_moons))

x_train, y_train = data_moons[:num_train, :-1], data_moons[:num_train, -1]
x_test, y_test = data_moons[num_test:, :-1], data_moons[num_test:, -1]

num_iterations = 20000
learning_rate = 0.1
n_h = 4
parameters_final = NN_model(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict(parameters_final, x_test.T)
Y_predictions_train = predict(parameters_final, x_train.T)


print(f"Train accuracy Moons: {np.mean(y_train == Y_predictions_train)}")
print(f"Test accuracy Moons: {np.mean(y_test == Y_predictions_test)}")

In [ ]:
data_blobs = np.array(df2)

num_train = int(.70 * len(data_blobs))
num_test = int(0.15 * len(data_blobs))

x_train, y_train = data_blobs[:num_train, :-1], data_blobs[:num_train, -1]
x_test, y_test = data_blobs[num_test:, :-1], data_blobs[num_test:, -1]

num_iterations = 20000
learning_rate = 0.1
n_h = 4
parameters_final = NN_model(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict(parameters_final, x_test.T)
Y_predictions_train = predict(parameters_final, x_train.T)


print(f"Train accuracy Blobs: {np.mean(y_train == Y_predictions_train)}")
print(f"Test accuracy Blobs: {np.mean(y_test == Y_predictions_test)}")

# Part 4

In [6]:
# This function taken from the CIFAR website

def unpickle(file):
    import pickle
    with open(file, 'rb') as fo:
        dict = pickle.load(fo, encoding='bytes')
    return dict

# Loaded in this way, each of the batch files contains a dictionary with the following elements:
#   data -- a 10000x3072 numpy array of uint8s. Each row of the array stores a 32x32 colour image. 
#           The first 1024 entries contain the red channel values, the next 1024 the green, and the final 1024 the blue. 
#           The image is stored in row-major order, so that the first 32 entries of the array are the red channel values 
#           of the first row of the image.
#   labels -- a list of 10000 numbers in the range 0-9. 
#             The number at index i indicates the label of the ith image in the array data.

def loadbatch(batchname):
    folder = 'cifar-10-batches-py'
    batch = unpickle(folder+"/"+batchname)
    return batch

def loadlabelnames():
    folder = 'cifar-10-batches-py'
    meta = unpickle(folder+"/"+'batches.meta')
    return meta[b'label_names']

batch1 = loadbatch('data_batch_1')
print("Number of items in the batch is", len(batch1))

# Display all keys, so we can see the ones we want
print('All keys in the batch:', batch1.keys())

data_cifar = batch1[b'data']
labels_cifar = batch1[b'labels']
print ("size of data in this batch:", len(data_cifar), ", size of labels:", len(labels_cifar))
print (type(data_cifar))
print(data_cifar.shape)

names = loadlabelnames()

Number of items in the batch is 4
All keys in the batch: dict_keys([b'batch_label', b'labels', b'data', b'filenames'])
size of data in this batch: 10000 , size of labels: 10000
<class 'numpy.ndarray'>
(10000, 3072)


In [7]:
# print(data)
data_new = data_cifar[1:100]
print(data_new[1:10])
print(data_new.shape)

data_new = data_new[:,:1023]

print(data_new.shape)
print(data_new[1:5])

num_train = int(.70 * len(data_new))
num_test = int(0.15 * len(data_new))

x_train, y_train = data_new[:num_train, :-1], data_new[:num_train, -1]
x_test, y_test = data_new[num_test:, :-1], data_new[num_test:, -1]

print(x_train)

[[255 253 253 ...  83  83  84]
 [ 28  37  38 ...  28  37  46]
 [170 168 177 ...  82  78  80]
 ...
 [134 131 128 ... 136 137 138]
 [125 110 102 ...  82  84  86]
 [ 53  54  56 ...  39  46  41]]
(99, 3072)
(99, 1023)
[[255 253 253 ...  68  78  79]
 [ 28  37  38 ...  67  54  63]
 [170 168 177 ...  71  75  71]
 [159 150 153 ... 157 166 173]]
[[154 126 105 ... 134 140 143]
 [255 253 253 ...  58  68  78]
 [ 28  37  38 ...  84  67  54]
 ...
 [221 214 216 ... 143 140 138]
 [110 103 104 ...  75  91  98]
 [151 150 151 ...  99 101 100]]


In [8]:
num_iterations = 20000
learning_rate = 0.1
n_h = 4
parameters_final = NN_model(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict(parameters_final, x_test.T)
Y_predictions_train = predict(parameters_final, x_train.T)

#print("train accuracy: {} %".format(100 - np.mean(np.abs(Y_predictions_train - y_train)) * 100))
print(f"Train accuracy: {np.mean(y_train == Y_predictions_train)}")
print(f"Test accuracy: {np.mean(y_test == Y_predictions_test)}")

#print("test accuracy: {} %".format(100 - np.mean(np.abs(Y_predictions_test - y_test)) * 100))

Train accuracy: 0.0
Test accuracy: 0.0


In [ ]:
def unpickle(file):
    import pickle
    with open(file, 'rb') as fo:
        dict = pickle.load(fo, encoding='bytes')
    return dict

def loadbatch(batchname):
    batch = unpickle('cifar-10-batches-py'+"/"+batchname)
    return batch

def loadlabelnames():
    meta = unpickle('cifar-10-batches-py'+"/"+'batches.meta')
    return meta[b'label_names']

batch1 = loadbatch('data_batch_1')
print("Number of items in the batch is", len(batch1))

# Display all keys, so we can see the ones we want
print('All keys in the batch:', batch1.keys())

data = batch1[b'data'][:,:1023]
labels = batch1[b'labels']
print ("size of data in this batch:", len(data), ", size of labels:", len(labels))
print (type(data))
# print(data.shape)
# print(labels)
label = []
data_class = []
names = loadlabelnames()

for i in range(len(labels)):
  if labels[i] == 1 or labels[i]==6:
    label.append(labels[i])
    data_class.append(data[i])

df = np.array(data_class)
df2 = np.array(label)


num_train = int(.70 * len(df))
num_test = int(0.15 * len(df))

x_train, y_train = data_class[:num_train], label[:num_train]
x_test, y_test = data_class[num_test:], label[num_test:]

x_train  = np.array(x_train)
y_train  = np.array(y_train)
x_test  = np.array(x_test)
y_test  = np.array(y_test)

arr = [y_train.tolist()]
y_train = np.array(arr)

arr1 = [y_test.tolist()]
y_test = np.array(arr1)

num_iterations = 20000
learning_rate = 0.01
n_h = 4
parameters_final = NN_model(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict(parameters_final, x_test.T)
Y_predictions_train = predict(parameters_final, x_train.T)


acc = np.mean(y_train == Y_predictions_train)
acc

### Part 5

### Tapan Auti

### Resources:

1. https://towardsdatascience.com/how-to-implement-an-adam-optimizer-from-scratch-76e7b217f1cc
2. https://towardsdatascience.com/adam-latest-trends-in-deep-learning-optimization-6be9a291375c
3. https://machinelearningmastery.com/tour-of-optimization-algorithms/

In [ ]:
def initialize_parameters(n_x, n_h, n_y):

    W1 = np.random.randn(n_h, n_x) * 0.01
    b1 = np.zeros(shape=(n_h, 1))
    W2 = np.random.randn(n_y, n_h) * 0.01
    b2 = np.zeros(shape=(n_y, 1))
    
    assert (W1.shape == (n_h, n_x))
    assert (b1.shape == (n_h, 1))
    assert (W2.shape == (n_y, n_h))
    assert (b2.shape == (n_y, 1))
    
    parameters = {"W1": W1,
                  "b1": b1,
                  "W2": W2,
                  "b2": b2}
    
    return parameters

def forward_pass(X, parameters):
    #print(parameters)
    # to make forward pass calculations we need W1 and W2 so we will extract them from dictionary parameters
    W1 = parameters['W1']
    W2 = parameters['W2']
    b1 = parameters['b1']
    b2 = parameters['b2']
    #print(parameters)
    #print(W1)
    #print(b1)
    # first layer calculations - hidden layer calculations
    Z1 = np.dot(W1, X) + b1
    #print(Z1)
    # for i in range(len(Z1)):
    #   print(i)
    #   print(Z1[i])
    A1 = tanh(Z1)  # activation in the first layer is tanh
    #print(A1)
    # output layer calculations
    Z2 = np.dot(W2, A1) + b2
    A2 = sigmoid(Z2)# A2 are predictions, y_hat
    
    # cache values for backpropagation calculations
    cache = {'Z1':Z1,
             'A1':A1,
             'Z2':Z2,
             'A2':A2
            }
    # print(A2.shape)
    return A2, cache

def backward_pass(parameters, cache, X, Y):
    
    # unpack paramaeters and cache to get values for backpropagation calculations
    W1 = parameters['W1']
    W2 = parameters['W2']
    
    Z1 = cache['Z1']
    A1 = cache['A1']
    Z2 = cache['Z2']
    A2 = cache['A2']
    
    m = X.shape[1] # number of examples in a training set
    #print(A2)
    #print(A2.shape)
    dZ2= A2 - Y
    
    
    dW2 = (1 / m) * np.dot(dZ2, A1.T)
    db2 = (1 / m) * np.sum(dZ2, axis=1, keepdims=True)   # keepdims - prevents python to output rank 1 array (n,)
    
    dZ1 = np.multiply(np.dot(W2.T, dZ2), 1 - np.power(A1, 2)) # we use tanh activation function
    dW1 = (1 / m) * np.dot(dZ1, X.T)  
    db1 = (1 / m) * np.sum(dZ1, axis=1, keepdims=True)
    
    grads = {"dW1": dW1,
             "db1": db1,
             "dW2": dW2,
             "db2": db2}
    
    return grads

def initialize_adam(parameters) :
       
    L = len(parameters) // 2 # number of layers in the neural networks
    v = {}
    s = {}
    
    # v- exponentially weighted average of the gradient
    # s -exponentially weighted average of the squared gradient
    for l in range(L):
        v["dW1" ] = np.zeros(parameters["W1" ].shape)
        v["db1" ] = np.zeros(parameters["b1" ].shape)
        s["dW2" ] = np.zeros(parameters["W2" ].shape)
        s["db2" ] = np.zeros(parameters["b2" ].shape)
        
    return v, s

def update_parameters_with_adam(parameters, grads, v, s, t, learning_rate = 0.01,
                                beta1 = 0.9, beta2 = 0.999,  epsilon = 1e-8):
   
  
    
    L = len(parameters) // 2                 # number of layers in the neural networks
    v_corrected = {}                         # Initializing first moment estimate, python dictionary
    s_corrected = {}                         # Initializing second moment estimate, python dictionary
    
        
    for l in range(L):
        v["dW1"] = beta1*v["dW1" ] + (1-beta1)*grads['dW1' ] 
        v["db1" ] = beta1*v["db1" ] + (1-beta1)*grads['db1' ]
       

        # Compute bias-corrected first moment estimate. 
       
        v_corrected["dW1" ] = v["dW1" ]/(1 - beta1**t)
        v_corrected["db1" ] = v["db1" ]/(1 - beta1**t)


        # Moving average of the squared gradients.
     
        s["dW2"] = beta2*s["dW2" ] + (1-beta2)*(grads['dW2' ])**2
        s["db2" ] = beta2*s["db2" ] + (1-beta2)*(grads['db2' ])**2
       

        # Compute bias-corrected second raw moment estimate. 
       
        s_corrected["dW2" ] = s["dW2" ]/(1 - beta2**t)
        s_corrected["db2" ] = s["db2" ]/(1 - beta2**t)
        

        parameters["W1" ] = parameters["W1" ]- (learning_rate*v_corrected["dW1" ])/(np.sqrt(s_corrected["dW2" ].T+epsilon))
        parameters["b1" ] = parameters["b1" ]- (learning_rate*v_corrected["db1" ])/(np.sqrt(s_corrected["db2" ].T+epsilon))
   
    return parameters, v, s

def NN_model(X,Y,n_h, num_iterations, learning_rate):
    
    n_x = X.shape[0] # size of an input layer = number of features 
    n_y = Y.shape[0] # size of an output layer
    parameters = initialize_parameters(n_x, n_h, n_y)
    
    #unpack parameters
    W1 = parameters["W1"]
    b1 = parameters["b1"]
    W2 = parameters["W2"]
    b2 = parameters["b2"]
    #print(parameters)
    param = {}
    for i in range(num_iterations):
      #print(i)  
      A2, cache = forward_pass(X, parameters)
      grads = backward_pass(parameters, cache, X, Y)
      v,s = initialize_adam(parameters)
      param = update_parameters_with_adam(parameters,grads,v,s , t = 2 )
      #print(i,param)
    #print(A2)
    # print(cache)    
    return parameters

def predict(parameters, X):
    
    A2, cache = forward_pass(X, parameters)
    predictions = np.round(A2)
    
    return predictions

def unpickle(file):
    import pickle
    with open(file, 'rb') as fo:
        dict = pickle.load(fo, encoding='bytes')
    return dict

def loadbatch(batchname):
    batch = unpickle('cifar-10-batches-py'+"/"+batchname)
    return batch

def loadlabelnames():
    meta = unpickle('cifar-10-batches-py'+"/"+'batches.meta')
    return meta[b'label_names']

batch1 = loadbatch('data_batch_1')
print("Number of items in the batch is", len(batch1))

# Display all keys, so we can see the ones we want
print('All keys in the batch:', batch1.keys())

data = batch1[b'data'][:,:1023]
labels = batch1[b'labels']
print ("size of data in this batch:", len(data), ", size of labels:", len(labels))
print (type(data))
# print(data.shape)
# print(labels)
label = []
data_class = []
names = loadlabelnames()

for i in range(len(labels)):
  if labels[i] == 1 or labels[i]==6:
    label.append(labels[i])
    data_class.append(data[i])

df = np.array(data_class)
df2 = np.array(label)


num_train = int(.70 * len(df))
num_test = int(0.15 * len(df))

x_train, y_train = data_class[:num_train], label[:num_train]
x_test, y_test = data_class[num_test:], label[num_test:]

x_train  = np.array(x_train)
y_train  = np.array(y_train)
x_test  = np.array(x_test)
y_test  = np.array(y_test)

arr = [y_train.tolist()]
y_train = np.array(arr)

arr1 = [y_test.tolist()]
y_test = np.array(arr1)

num_iterations = 20000
learning_rate = 0.01
n_h = 4
parameters_final = NN_model(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict(parameters_final, x_test.T)
Y_predictions_train = predict(parameters_final, x_train.T)


acc = np.mean(y_train == Y_predictions_train)
acc

# Part 5 : Atharva Kulkarni

Changed the activation function from a standard sigmoid to ReLU:


In [ ]:
def forward_pass_relu(X, parameters):
    
    # to make forward pass calculations we need W1 and W2 so we will extract them from dictionary parameters
    W1 = parameters['W1']
    W2 = parameters['W2']
    b1 = parameters['b1']
    b2 = parameters['b2']

    # first layer calculations - hidden layer calculations
    Z1 = np.dot(W1,X) + b1
    A1 = relu(Z1)  # activation in the first layer is tanh
    
    # output layer calculations
    Z2 = np.dot(W2, A1) + b2
    A2 = relu(Z2)# A2 are predictions, y_hat
    
    # cache values for backpropagation calculations
    cache = {'Z1':Z1,
             'A1':A1,
             'Z2':Z2,
             'A2':A2
            }
    # print(A2.shape)
    return A2, cache

def NN_model_relu(X,Y,n_h, num_iterations, learning_rate):
    
    n_x = X.shape[0] # size of an input layer = number of features 
    n_y = Y.shape[0] # size of an output layer
    parameters = initialize_parameters(n_x, n_h, n_y)
    
    #unpack parameters
    W1 = parameters["W1"]
    b1 = parameters["b1"]
    W2 = parameters["W2"]
    b2 = parameters["b2"]
    
    for i in range(num_iterations):
        A2, cache = forward_pass_relu(X, parameters)
        grads = backward_pass(parameters, cache, X, Y)
        parameters = update_parameters(parameters, learning_rate, grads)
        
    return parameters

def predict_relu(parameters, X):
    
    A2, cache = forward_pass_relu(X, parameters)
    predictions = np.round(A2)
    
    return predictions

In [ ]:
num_iterations = 20000
learning_rate = 0.01
n_h = 4
parameters_final = NN_model_relu(x_train.T,y_train,n_h, num_iterations, learning_rate)

Y_predictions_test = predict_relu(parameters_final, x_test.T)
Y_predictions_train = predict_relu(parameters_final, x_train.T)


acc = np.mean(y_train == Y_predictions_train)
acc